### Part 1: Scrape from the Wikipedia Page >> https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M

In [187]:
import requests
import urllib.request
import time
from bs4 import BeautifulSoup
import numpy as np
import pandas as pd
from urllib.request import urlopen

url = "https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M"
html = urlopen(url)

soup = BeautifulSoup(html, "html.parser")
soup_table = soup.find_all('table')

In [188]:
postal_code = []
borough = []
neighborhood = []

for table in soup_table:
    rows = table.find_all('tr')
    
    for row in rows:
        cells = row.find_all('td')
        if len(cells) == 3:
            temp = cells[0]
            postal_code.append(temp.text.strip())
                        
            temp = cells[1]
            borough.append(temp.text.strip())
            
            temp = cells[2]
            neighborhood.append(temp.text.strip())
            

###### For rows with column "borough" "Not assigned", remove them.

In [189]:
df = pd.DataFrame()
df["postal_code"] = postal_code
df["borough"] = borough
df["neighborhood"] = neighborhood
df = df[df["borough"]!="Not assigned"]

###### For rows with column "neighborhood" "Not assigned", assign it the same value as column "borough"

In [190]:
df[df["neighborhood"] == "Not assigned"].neighborhood = df["borough"]

In [191]:
df.head()

postal_code           borough                                 neighborhood
2         M3A        North York                                    Parkwoods
3         M4A        North York                             Victoria Village
4         M5A  Downtown Toronto                    Regent Park, Harbourfront
5         M6A        North York             Lawrence Manor, Lawrence Heights
6         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government

In [192]:
df.shape

(103, 3)

###### There are in total 103 entries in the cleaned-up table.

---
---

### Part 2: Get the latitude and longitude coordinates of each neighborhood.

In [193]:
df_code = pd.read_csv("https://cocl.us/Geospatial_data")

df_code.head()

Postal Code   Latitude  Longitude
0         M1B  43.806686 -79.194353
1         M1C  43.784535 -79.160497
2         M1E  43.763573 -79.188711
3         M1G  43.770992 -79.216917
4         M1H  43.773136 -79.239476

In [194]:
df_wll = pd.merge(df, df_code, how='left', left_on = 'postal_code', right_on = 'Postal Code')
df_wll.drop("Postal Code", axis=1, inplace = True)

In [195]:
df_wll.head()

postal_code           borough                                 neighborhood  \
0         M3A        North York                                    Parkwoods   
1         M4A        North York                             Victoria Village   
2         M5A  Downtown Toronto                    Regent Park, Harbourfront   
3         M6A        North York             Lawrence Manor, Lawrence Heights   
4         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government   

    Latitude  Longitude  
0  43.753259 -79.329656  
1  43.725882 -79.315572  
2  43.654260 -79.360636  
3  43.718518 -79.464763  
4  43.662301 -79.389494

--- 
---

### Part 3: Explore and cluster the neighborhoods in Toronto.

###### Narrow down the Borough related to Toronto

In [196]:
toronto_data = df_wll[df_wll['borough'].str.contains('Toronto')].reset_index(drop=True)

In [197]:
toronto_data.shape

(39, 5)

In [198]:
print('The dataframe has {} boroughs and {} neighborhoods.'.format(
        len(toronto_data['borough'].unique()),
        toronto_data.shape[0]
    )
)

The dataframe has 4 boroughs and 39 neighborhoods.


###### Get the latitude & longitude of Toronto

In [202]:
from geopy.geocoders import Nominatim

address = 'Toronto, Ontario'

geolocator = Nominatim(user_agent="ny_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of Toronto are {}, {}.'.format(latitude, longitude))

#latitude = 43.651070
#longitude = -79.347015


###### Create a map of New York with neighborhoods superimposed on top

In [204]:
import folium

# create map of New York using latitude and longitude values
map_toronto = folium.Map(location=[latitude, longitude], zoom_start=10)

# add markers to map
for lat, lng, postal_code, borough, neighborhood in zip(toronto_data['Latitude'], toronto_data['Longitude'], toronto_data['postal_code'], toronto_data['borough'], toronto_data['neighborhood']):
    label = '{}, {}, {}'.format(postal_code, neighborhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_toronto)  
    
map_toronto

###### Config the libraries and credential for connecting to Four Square

In [279]:
import requests # library to handle requests
import json # library to handle JSON files

CLIENT_ID = '<hide>' # your Foursquare ID
CLIENT_SECRET = '<hide' # your Foursquare Secret
VERSION = '20180605' # Foursquare API version
LIMIT = 100 # A default Foursquare API limit value



###### Explore Neighborhoods in Toronto

In [207]:
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

###### Get the nearby venues for all the neighborhoods

In [208]:
toronto_venues = getNearbyVenues(names=toronto_data['neighborhood'],
                                   latitudes=toronto_data['Latitude'],
                                   longitudes=toronto_data['Longitude']
                                  )

Regent Park, Harbourfront
Queen's Park, Ontario Provincial Government
Garden District, Ryerson
St. James Town
The Beaches
Berczy Park
Central Bay Street
Christie
Richmond, Adelaide, King
Dufferin, Dovercourt Village
Harbourfront East, Union Station, Toronto Islands
Little Portugal, Trinity
The Danforth West, Riverdale
Toronto Dominion Centre, Design Exchange
Brockton, Parkdale Village, Exhibition Place
India Bazaar, The Beaches West
Commerce Court, Victoria Hotel
Studio District
Lawrence Park
Roselawn
Davisville North
Forest Hill North & West, Forest Hill Road Park
High Park, The Junction South
North Toronto West,  Lawrence Park
The Annex, North Midtown, Yorkville
Parkdale, Roncesvalles
Davisville
University of Toronto, Harbord
Runnymede, Swansea
Moore Park, Summerhill East
Kensington Market, Chinatown, Grange Park
Summerhill West, Rathnelly, South Hill, Forest Hill SE, Deer Park
CN Tower, King and Spadina, Railway Lands, Harbourfront West, Bathurst Quay, South Niagara, Island airport


In [211]:
print('There are {} venues records amongs all the neighborhoods.'.format(toronto_venues.shape[0]))
toronto_venues.head()

There are 1624 venues records amongs all the neighborhoods.


Neighborhood  Neighborhood Latitude  Neighborhood Longitude  \
0  Regent Park, Harbourfront               43.65426              -79.360636   
1  Regent Park, Harbourfront               43.65426              -79.360636   
2  Regent Park, Harbourfront               43.65426              -79.360636   
3  Regent Park, Harbourfront               43.65426              -79.360636   
4  Regent Park, Harbourfront               43.65426              -79.360636   

                    Venue  Venue Latitude  Venue Longitude  \
0        Roselle Desserts       43.653447       -79.362017   
1           Tandem Coffee       43.653559       -79.361809   
2  Cooper Koo Family YMCA       43.653249       -79.358008   
3     Body Blitz Spa East       43.654735       -79.359874   
4          Impact Kitchen       43.656369       -79.356980   

        Venue Category  
0               Bakery  
1          Coffee Shop  
2  Distribution Center  
3                  Spa  
4           Restaurant

###### Number of venues were returned for each neighborhood

In [212]:
toronto_venues.groupby('Neighborhood').count()

Neighborhood Latitude  \
Neighborhood                                                                
Berczy Park                                                            55   
Brockton, Parkdale Village, Exhibition Place                           23   
Business reply mail Processing Centre, South Ce...                     16   
CN Tower, King and Spadina, Railway Lands, Harb...                     16   
Central Bay Street                                                     68   
Christie                                                               16   
Church and Wellesley                                                   75   
Commerce Court, Victoria Hotel                                        100   
Davisville                                                             33   
Davisville North                                                        9   
Dufferin, Dovercourt Village                                           13   
First Canadian Place, Underground city                                100   
Forest Hill North & West, Forest Hill Road Park                         4   
Garden District, Ryerson                                              100   
Harbourfront East, Union Station, Toronto Islands                     100   
High Park, The Junction South                                          25   
India Bazaar, The Beaches West                                         19   
Kensington Market, Chinatown, Grange Park                              74   
Lawrence Park                                                           3   
Little Portugal, Trinity                                               45   
Moore Park, Summerhill East                                             2   
North Toronto West,  Lawrence Park                                     18   
Parkdale, Roncesvalles                                                 14   
Queen's Park, Ontario Provincial Government                            33   
Regent Park, Harbourfront                                              44   
Richmond, Adelaide, King                                              100   
Rosedale                                                                4   
Roselawn                                                                2   
Runnymede, Swansea                                                     33   
St. James Town                                                         85   
St. James Town, Cabbagetown                                            48   
Stn A PO Boxes                                                         96   
Studio District                                                        37   
Summerhill West, Rathnelly, South Hill, Forest ...                     14   
The Annex, North Midtown, Yorkville                                    19   
The Beaches                                                             4   
The Danforth West, Riverdale                                           43   
Toronto Dominion Centre, Design Exchange                              100   
University of Toronto, Harbord                                         34   

                                                    Neighborhood Longitude  \
Neighborhood                                                                 
Berczy Park                                                             55   
Brockton, Parkdale Village, Exhibition Place                            23   
Business reply mail Processing Centre, South Ce...                      16   
CN Tower, King and Spadina, Railway Lands, Harb...                      16   
Central Bay Street                                                      68   
Christie                                                                16   
Church and Wellesley                                                    75   
Commerce Court, Victoria Hotel                                         100   
Davisville                                                              33   
Davisville North                    

In [213]:
print('There are {} uniques categories.'.format(len(toronto_venues['Venue Category'].unique())))

There are 235 uniques categories.


###### Analyze each neighborhood

In [217]:
# one hot encoding
toronto_onehot = pd.get_dummies(toronto_venues[['Venue Category']], prefix="", prefix_sep="")
toronto_onehot["Neighbor"] = toronto_venues["Neighborhood"]

fixed_columns = [toronto_onehot.columns[-1]] + list(toronto_onehot.columns[:-1])
toronto_onehot = toronto_onehot[fixed_columns]

toronto_onehot

Neighbor  Afghan Restaurant  \
0                             Regent Park, Harbourfront                  0   
1                             Regent Park, Harbourfront                  0   
2                             Regent Park, Harbourfront                  0   
3                             Regent Park, Harbourfront                  0   
4                             Regent Park, Harbourfront                  0   
...                                                 ...                ...   
1619  Business reply mail Processing Centre, South C...                  0   
1620  Business reply mail Processing Centre, South C...                  0   
1621  Business reply mail Processing Centre, South C...                  0   
1622  Business reply mail Processing Centre, South C...                  0   
1623  Business reply mail Processing Centre, South C...                  0   

      Airport  Airport Food Court  Airport Gate  Airport Lounge  \
0           0                   0             0               0   
1           0                   0             0               0   
2           0                   0             0               0   
3           0                   0             0               0   
4           0                   0             0               0   
...       ...                 ...           ...             ...   
1619        0                   0             0               0   
1620        0                   0             0               0   
1621        0                   0             0               0   
1622        0                   0             0               0   
1623        0                   0             0               0   

      Airport Service  Airport Terminal  American Restaurant  Antique Shop  \
0                   0                 0                    0             0   
1                   0                 0                    0             0   
2                   0                 0                    0             0   
3                   0                 0                    0             0   
4                   0                 0                    0             0   
...               ...               ...                  ...           ...   
1619                0                 0                    0             0   
1620                0                 0                    0             0   
1621                0                 0                    0             0   
1622                0                 0                    0             0   
1623                0                 0                    0             0   

      ...  Theater  Theme Restaurant  Toy / Game Store  Trail  Train Station  \
0     ...        0                 0                 0      0              0   
1     ...        0                 0                 0      0              0   
2     ...        0                 0                 0      0              0   
3     ...        0                 0                 0      0              0   
4     ...        0                 0                 0      0              0   
...   ...      ...               ...               ...    ...            ...   
1619  ...        0                 0                 0      0              0   
1620  ...        0                 0                 0      0              0   
1621  ...        0                 0                 0      0              0   
1622  ...        0                 0                 0      0              0   
1623  ...        0                 0                 0      0              0   

      Vegetarian / Vegan Restaurant  Video Game Store  Vietnamese Restaurant  \
0                                 0                 0                      0   
1                                 0                 0                      0   
2                                 0                 0                      0   
3                                 0                 0                      0   
4      

###### Group rows by neighborhood and by taking the mean of the frequency of occurrence of each category

In [245]:
toronto_grouped = toronto_onehot.groupby('Neighbor').mean().reset_index()
toronto_grouped

Neighbor  Afghan Restaurant  \
0                                         Berczy Park           0.000000   
1        Brockton, Parkdale Village, Exhibition Place           0.000000   
2   Business reply mail Processing Centre, South C...           0.000000   
3   CN Tower, King and Spadina, Railway Lands, Har...           0.000000   
4                                  Central Bay Street           0.000000   
5                                            Christie           0.000000   
6                                Church and Wellesley           0.013333   
7                      Commerce Court, Victoria Hotel           0.000000   
8                                          Davisville           0.000000   
9                                    Davisville North           0.000000   
10                       Dufferin, Dovercourt Village           0.000000   
11             First Canadian Place, Underground city           0.000000   
12    Forest Hill North & West, Forest Hill Road Park           0.000000   
13                           Garden District, Ryerson           0.000000   
14  Harbourfront East, Union Station, Toronto Islands           0.000000   
15                      High Park, The Junction South           0.000000   
16                     India Bazaar, The Beaches West           0.000000   
17          Kensington Market, Chinatown, Grange Park           0.000000   
18                                      Lawrence Park           0.000000   
19                           Little Portugal, Trinity           0.000000   
20                        Moore Park, Summerhill East           0.000000   
21                 North Toronto West,  Lawrence Park           0.000000   
22                             Parkdale, Roncesvalles           0.000000   
23        Queen's Park, Ontario Provincial Government           0.000000   
24                          Regent Park, Harbourfront           0.000000   
25                           Richmond, Adelaide, King           0.000000   
26                                           Rosedale           0.000000   
27                                           Roselawn           0.000000   
28                                 Runnymede, Swansea           0.000000   
29                                     St. James Town           0.000000   
30                        St. James Town, Cabbagetown           0.000000   
31                                     Stn A PO Boxes           0.000000   
32                                    Studio District           0.000000   
33  Summerhill West, Rathnelly, South Hill, Forest...           0.000000   
34                The Annex, North Midtown, Yorkville           0.000000   
35                                        The Beaches           0.000000   
36                       The Danforth West, Riverdale           0.000000   
37           Toronto Dominion Centre, Design Exchange           0.000000   
38                     University of Toronto, Harbord           0.000000   

    Airport  Airport Food Court  Airport Gate  Airport Lounge  \
0    0.0000              0.0000        0.0000           0.000   
1    0.0000              0.0000        0.0000           0.000   
2    0.0000              0.0000        0.0000           0.000   
3    0.0625              0.0625        0.0625           0.125   
4    0.0000              0.0000        0.0000           0.000   
5    0.0000              0.0000        0.0000           0.000   
6    0.0000              0.0000        0.0000           0.000   
7    0.0000              0.0000        0.0000           0.000   
8    0.0000              0.0000        0.0000           0.000   
9    0.0000              0.0000        0.0000           0.000   
10   0.0000              0.0000        0.0000           0.000   
11   0.0000              0.0000        0.0000           0.000   
12   0.0000              0.0000        0.0000           0.000   
13   0.0000              0.0000        0.0000           0.000   
14   0.0000              0.00

###### Create the new dataframe and display the top 10 venues for each neighborhood.

In [220]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

In [221]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighborhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighborhood'] = toronto_grouped['Neighbor']

for ind in np.arange(toronto_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(toronto_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.head()

Neighborhood 1st Most Common Venue  \
0                                        Berczy Park           Coffee Shop   
1       Brockton, Parkdale Village, Exhibition Place                  Café   
2  Business reply mail Processing Centre, South C...            Skate Park   
3  CN Tower, King and Spadina, Railway Lands, Har...        Airport Lounge   
4                                 Central Bay Street           Coffee Shop   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0                Bakery          Cocktail Bar              Beer Bar   
1        Breakfast Spot             Nightclub           Coffee Shop   
2           Pizza Place               Brewery         Burrito Place   
3       Airport Service              Boutique       Harbor / Marina   
4                  Café    Italian Restaurant        Sandwich Place   

  5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0           Cheese Shop    Seafood Restaurant            Restaurant   
1          Climbing Gym         Burrito Place            Restaurant   
2            Restaurant        Farmers Market  Fast Food Restaurant   
3      Sculpture Garden         Boat or Ferry   Rental Car Location   
4   Japanese Restaurant       Bubble Tea Shop           Salad Place   

  8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
0        Farmers Market        Sandwich Place         Breakfast Spot  
1    Italian Restaurant          Intersection                    Bar  
2               Butcher      Recording Studio          Auto Workshop  
3                   Bar           Coffee Shop                  Plane  
4          Burger Joint      Department Store        Thai Restaurant

###### Run k-means to cluster the neighborhood into 5 clusters

In [246]:
# import k-means from clustering stage
from sklearn.cluster import KMeans

In [274]:
# set number of clusters
kclusters = 5

toronto_grouped_clustering = toronto_grouped.drop('Neighbor', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(toronto_grouped_clustering) 

In [275]:
toronto_grouped.shape

(39, 236)

In [276]:
kmeans.labels_

array([2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 0, 2, 2, 2, 2, 2, 0, 2, 1, 2,
       2, 2, 2, 2, 0, 4, 2, 2, 2, 2, 2, 2, 2, 3, 2, 2, 2])

###### Create a new dataframe that includes the cluster as well as the top 10 venues for each neighborhood

In [277]:
# add clustering labels
neighborhoods_venues_sorted.drop('Cluster Labels', axis=1, inplace= True)
neighborhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

toronto_merged = toronto_data

# merge manhattan_grouped with manhattan_data to add latitude/longitude for each neighborhood
toronto_merged = pd.merge(toronto_data, neighborhoods_venues_sorted, how="inner", left_on = "neighborhood", right_on = "Neighborhood")

toronto_merged.head() # check the last columns!

postal_code           borough                                 neighborhood  \
0         M5A  Downtown Toronto                    Regent Park, Harbourfront   
1         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government   
2         M5B  Downtown Toronto                     Garden District, Ryerson   
3         M5C  Downtown Toronto                               St. James Town   
4         M4E      East Toronto                                  The Beaches   

    Latitude  Longitude  Cluster Labels  \
0  43.654260 -79.360636               2   
1  43.662301 -79.389494               2   
2  43.657162 -79.378937               2   
3  43.651494 -79.375418               2   
4  43.676357 -79.293031               3   

                                  Neighborhood 1st Most Common Venue  \
0                    Regent Park, Harbourfront           Coffee Shop   
1  Queen's Park, Ontario Provincial Government           Coffee Shop   
2                     Garden District, Ryerson        Clothing Store   
3                               St. James Town           Coffee Shop   
4                                  The Beaches                   Pub   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0                Bakery                  Park                   Pub   
1           Yoga Studio                 Diner            Restaurant   
2           Coffee Shop                  Café   Japanese Restaurant   
3                  Café          Cocktail Bar            Restaurant   
4                 Trail     Health Food Store          Neighborhood   

   5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0                   Café               Theater        Breakfast Spot   
1  Portuguese Restaurant                  Park           Music Venue   
2        Bubble Tea Shop        Cosmetics Shop                 Diner   
3               Beer Bar             Gastropub   American Restaurant   
4            Yoga Studio               Dog Run    Dim Sum Restaurant   

  8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
0           Event Space            Shoe Store                  Hotel  
1    Mexican Restaurant    Italian Restaurant             Hobby Shop  
2        Lingerie Store      Ramen Restaurant     Italian Restaurant  
3        Farmers Market                 Hotel    Japanese Restaurant  
4                 Diner        Discount Store    Distribution Center

###### Visualize the resulting clusters

In [278]:
# Matplotlib and associated plotting modules
import matplotlib.cm as cm
import matplotlib.colors as colors

# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(toronto_merged['Latitude'], toronto_merged['Longitude'], toronto_merged['Neighborhood'], toronto_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

###### Assign the Name for each cluster

###### Cluster 0: Park Areas

In [255]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 0, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

borough  Cluster Labels  \
18   Central Toronto               0   
21   Central Toronto               0   
33  Downtown Toronto               0   

                                       Neighborhood 1st Most Common Venue  \
18                                    Lawrence Park                  Park   
21  Forest Hill North & West, Forest Hill Road Park                  Park   
33                                         Rosedale                  Park   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
18              Bus Line           Swim School  Colombian Restaurant   
21      Sushi Restaurant         Jewelry Store                 Trail   
33                 Trail            Playground         Deli / Bodega   

   5th Most Common Venue        6th Most Common Venue  \
18          Dessert Shop            Electronics Store   
21     Electronics Store  Eastern European Restaurant   
33     Electronics Store  Eastern European Restaurant   

          7th Most Common Venue 8th Most Common Venue 9th Most Common Venue  \
18  Eastern European Restaurant   Dumpling Restaurant            Donut Shop   
21          Dumpling Restaurant            Donut Shop      Doner Restaurant   
33          Dumpling Restaurant            Donut Shop      Doner Restaurant   

   10th Most Common Venue  
18       Doner Restaurant  
21                Dog Run  
33                Dog Run

###### Cluster 1: Trail Areas

In [256]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 1, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

borough  Cluster Labels                 Neighborhood  \
29  Central Toronto               1  Moore Park, Summerhill East   

   1st Most Common Venue 2nd Most Common Venue 3rd Most Common Venue  \
29                 Trail            Playground           Yoga Studio   

   4th Most Common Venue 5th Most Common Venue        6th Most Common Venue  \
29         Deli / Bodega     Electronics Store  Eastern European Restaurant   

   7th Most Common Venue 8th Most Common Venue 9th Most Common Venue  \
29   Dumpling Restaurant            Donut Shop      Doner Restaurant   

   10th Most Common Venue  
29                Dog Run

###### Cluster 2: Dining Areas

In [257]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 2, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

borough  Cluster Labels  \
0   Downtown Toronto               2   
1   Downtown Toronto               2   
2   Downtown Toronto               2   
3   Downtown Toronto               2   
5   Downtown Toronto               2   
6   Downtown Toronto               2   
7   Downtown Toronto               2   
8   Downtown Toronto               2   
9       West Toronto               2   
10  Downtown Toronto               2   
11      West Toronto               2   
12      East Toronto               2   
13  Downtown Toronto               2   
14      West Toronto               2   
15      East Toronto               2   
16  Downtown Toronto               2   
17      East Toronto               2   
20   Central Toronto               2   
22      West Toronto               2   
23   Central Toronto               2   
24   Central Toronto               2   
25      West Toronto               2   
26   Central Toronto               2   
27  Downtown Toronto               2   
28      West Toronto               2   
30  Downtown Toronto               2   
31   Central Toronto               2   
32  Downtown Toronto               2   
34  Downtown Toronto               2   
35  Downtown Toronto               2   
36  Downtown Toronto               2   
37  Downtown Toronto               2   
38      East Toronto               2   

                                         Neighborhood 1st Most Common Venue  \
0                           Regent Park, Harbourfront           Coffee Shop   
1         Queen's Park, Ontario Provincial Government           Coffee Shop   
2                            Garden District, Ryerson        Clothing Store   
3                                      St. James Town           Coffee Shop   
5                                         Berczy Park           Coffee Shop   
6                                  Central Bay Street           Coffee Shop   
7                                            Christie         Grocery Store   
8                            Richmond, Adelaide, King           Coffee Shop   
9                        Dufferin, Dovercourt Village              Pharmacy   
10  Harbourfront East, Union Station, Toronto Islands           Coffee Shop   
11                           Little Portugal, Trinity                   Bar   
12                       The Danforth West, Riverdale      Greek Restaurant   
13           Toronto Dominion Centre, Design Exchange           Coffee Shop   
14       Brockton, Parkdale Village, Exhibition Place                  Café   
15                     India Bazaar, The Beaches West                  Park   
16                     Commerce Court, Victoria Hotel           Coffee Shop   
17                                    Studio District           Coffee Shop   
20                                   Davisville North                  Park   
22                      High Park, The Junction South                  Café   
23                 North Toronto West,  Lawrence Park           Coffee Shop   
24                The Annex, North Midtown, Yorkville        Sandwich Place   
25                             Parkdale, Roncesvalles        Breakfast Spot   
26                                         Davisville        Sandwich Place   
27                     University of Toronto, Harbord                  Café   
28                                 Runnymede, Swansea                  Café   
30          Kensington Market, Chinatown, Grange Park    Mexican Restaurant   
31  Summerhill West, Rathnelly, South Hill, Forest...           Coffee Shop   
32  CN Tower, King and Spadina, Railway Lands, Har...        Airport Lounge   
34                                     Stn A PO Boxes           Coffee Shop   
35                        St. James Town, Cabbagetown           Coffee Shop   
36             First Canadian Place, Underground city           Coffee Shop   
37                               Church and Wellesley           Coffee Shop   
38  Business reply mail Processing Centre, So

###### Cluster 3: Relaxing Areas

In [258]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 3, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

borough  Cluster Labels Neighborhood 1st Most Common Venue  \
4  East Toronto               3  The Beaches                   Pub   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
4                 Trail     Health Food Store          Neighborhood   

  5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
4           Yoga Studio               Dog Run    Dim Sum Restaurant   

  8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
4                 Diner        Discount Store    Distribution Center

###### Cluster 4: Others

In [259]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 4, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

borough  Cluster Labels Neighborhood 1st Most Common Venue  \
19  Central Toronto               4     Roselawn           Music Venue   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
19                Garden           Yoga Studio      Department Store   

   5th Most Common Venue 6th Most Common Venue        7th Most Common Venue  \
19           Escape Room     Electronics Store  Eastern European Restaurant   

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
19   Dumpling Restaurant            Donut Shop       Doner Restaurant